### ISSUE TITLE
Handling escape symbols

### ISSUE THREAD

**Issue by jonathantneal** [postcss/postcss#1349]
From https://acss.io/#pseudo-classes:

Each line should parse as one word (i.e. one identifier):

```css
C\(\#0280ae\)
C\(brandColor\)
C\(\#fff\)\:h:hover
```

This can be verified with https://rawgit.com/tabatkins/parse-css/master/example.html

This issue continues from https://github.com/shellscape/postcss-values-parser/issues/93

With thanks and credit to @nex3 and @ai for identifying this issue.

**Comment by ai** [comment]
1. What are CSS escaping rules? Ignore the next symbol after `\`?
2. What is the current output?

**Comment by jonathantneal** [comment]
> 1. What are CSS escaping rules? Ignore the next symbol after \?

The general escape rule is that a backslash swallows the next character ([Reference](https://drafts.csswg.org/css-syntax/#starts-with-a-valid-escape)).

Within names (or URLs), an escape cannot consume a newline, but within strings it can.

[![CSS Escape Diagram](https://user-images.githubusercontent.com/188426/79638848-c5dcb300-8155-11ea-88a8-edba9d65c1dc.png)](https://drafts.csswg.org/css-syntax/#escape-diagram)

In the above diagram, you’ll see exceptions for hex digits, but AFAIK we can ignore this, because escape codes always begin words (or what the spec calls "identifiers") and hex digits are already covered by identifier rules.

> 2. What is the current output?

Given the CSS `'f\:1`, the token output is this:

```
[
  [ 'word', 'f', 1, 1, 1, 2 ],
  [ 'word', '\\:', 1, 3, 1, 4 ],
  [ 'word', '1', 1, 5, 1, 5 ]
]
```

And it should be more like this:

```
[
  [ 'word', 'f\\:1', 1, 1, 1, 5 ]
]

**Comment by jonathantneal** [comment]
I only have a small update. I’m also wrapping up school with the kids and need to start work, so please forgive typos or confusing statements.

I was able to complete my first pass of the tokenizer this weekend. Here is the source:

https://gist.github.com/jonathantneal/dd859523df584b012dfa1cc8795d7ca9

Here are some benchmark results.

```
PostCSS Tokenizer 7.0.27     x 133 ops/sec ±1.69% (79 runs sampled)
PostCSS Tokenizer 2020-04-20 x 72.12 ops/sec ±3.51% (77 runs sampled)
[09:38:31] Fastest test is PostCSS at 1.85x faster than t20200420

PostCSS:   8 ms
t20200420: 14 ms (1.8 times slower)
```

```
[09:38:53]    PostCSS x 113 ops/sec ±5.00% (70 runs sampled)
[09:38:59]    t20200420 x 70.98 ops/sec ±3.91% (77 runs sampled)
[09:38:59] Fastest test is PostCSS at 1.59x faster than t20200420

PostCSS:   9 ms
t20200420: 14 ms (1.6 times slower)
```

😢 Yes, it is significantly slower.
🤞 I hope it is slower because I wrote it poorly.
🤔 I understand that a spec-compliant tokenizer should be as performant as the current tokenizer to even consider such a change.

And here is an example structure of a token:

```js
// where css is "-.567800E-0012780em"
const tokenExample1 = [
  /** @type {TokenType}  */ "number",
  /** @type {TokenValue} */ "-.567800E-0012780em",
  /** @type {TokenStart} */ 0,
  /** @type {TokenEnd}   */ 19,
  /** @type {TokenSplit} */ 17
]

// where css is "1138--thx"
const tokenExample2 = [
  /** @type {TokenType}  */ "number",
  /** @type {TokenValue} */ "1138--thx",
  /** @type {TokenStart} */ 0,
  /** @type {TokenEnd}   */ 9,
  /** @type {TokenSplit} */ 4
]

// where css is "foo\:bar"
const tokenExample3 = [
  /** @type {TokenType}  */ "identifier",
  /** @type {TokenValue} */ "foo\\:bar",
  /** @type {TokenStart} */ 0,
  /** @type {TokenEnd}   */ 8,
  /** @type {TokenSplit} */ 0
]

/** @typedef {"comment" | "delimiter" | "identifier" | "number" | "string" | "whitespace"} TokenType - Name of the token. */
/** @typedef {string} TokenValue - Value of the token. */
/** @typedef {number} TokenStart - Character index of the source where the token starts. */
/** @typedef {number} TokenEnd - Character index of the source where the token ends. */
/** @typedef {number} TokenSplit - Character index of the source where the token can be split. Splits occur between numbers and their units, and splits occur between identifier symbols (like `@` or `#`) and their identifier names). */
```

**Comment by nex3** [comment]
Thanks a ton for looking into this, @jonathantneal! Maybe @ai can provide some performance tips?

**Comment by ai** [comment]
Do we really need to change it in the tokenizer if it affects performance?

We can change parser and avoid any performance investigations.

**Comment by jonathantneal** [comment]
Having some time tonight, I’ve pushed the latest version of the tokenizer to https://github.com/csstools/tokenizer

It includes benchmark tests and the ability to create token JSON files, if that’s at all helpful.

These benchmarks were averaged from my local machine:

```
PostCSS Tokenizer 7.0.27:      49548 tokens in  7 ms
PostCSS Tokenizer Development: 60458 tokens in 14 ms (2.0 times slower)
```

---

> Do we really need to change it in the tokenizer if it affects performance?

I don’t know, because I don’t yet understand some parts of the current tokenizer. For instance, the current tokenizer produces a single token from CSS like `sans-serif,` while the tokenizer in development produces 2 tokens (for `sans-serif` and `,`).

I think the answer to your question has a couple dependencies. The first dependency is to know the performance impact the current parser adds to tokenization that might be unneeded with this other tokenizer. The second dependency is whether the current tokenizer and parser can be more surgically modified to produce accurate results without losing performance — at least with escapes.

> We can change parser and avoid any performance investigations.

Would you have time to investigate this? My earlier comment describes how escapes are assembled.

**Comment by jonathantneal** [comment]
I want to provide more frequent updates here, but my free time is split between this and a significant upgrade to PostCSS Preset Env.

I think the next thing I can do is plug the tokenizer into the PostCSS object model, and see what the final performance impact is on Bootstrap’s CSS.

Let me admit that I have trouble solving the small issue (escapes) without solving a big issue (css syntax compliance). I do believe solving the bigger issue will knock out the little issue, tho. My hope is that I can push this far enough along that you and others can bring it across the finish line.

**Comment by jonathantneal** [comment]
Please let me know if I should post less frequently. Here are my updates:

#### In Progress: Combine the Experimental Tokenizer with the PostCSS Object Model

I have not yet integrated the experimental tokenizer into the PostCSS object model. I think I need a highly focused 4 hour window to accomplish this; time I have not yet had. Once this is done, I should be able to verify whether the escape symbols are properly handled with this combination. The weekend is soon.

#### Now: Compression May Improve Performance

I have added a compression step to the experimental tokenizer using [terser](https://github.com/terser/terser), and the compressed tokenizer is regularly scoring better benchmarks than the uncompressed version. A unique element of the compression is that it is replacing every character code variable with the actual character code number; e.g. `case TAB` becomes `case 9`.

As of April 30, 2020, these benchmarks were averaged from my local machine:

```
PostCSS Tokenizer 7.0.27:            49548 tokens in 10 ms
PostCSS Tokenizer Development (min): 60458 tokens in 14 ms (1.4 times slower)
PostCSS Tokenizer Development:       60458 tokens in 18 ms (1.8 times slower)
```

Benchmarks are not an exact science for me. I have not changed anything about the experimental tokenizer, but it is more regularly parsing 1.8x slower, an improvement from 2x slower.  Still, the compressed experimental tokenizer is benchmarking only 1.4x slower than the current PostCSS tokenizer, and it is picking up 1.2x the tokens.

#### Soon: Improving Syntax Tests

I would like to migrate some of the [css-syntax tests from the web-platform-tests project](http://wpt.live/css/css-syntax/) into this development work. This is lower priority than integrating the Object Model, and I will fight the temptation to do this first. With new versions of these tests, PostCSS could protect itself against syntactically valid edge cases, and even trail-blaze work toward the [Houdini Parser API](https://github.com/WICG/CSS-Parser-API/blob/master/README.md). The WPT tests themselves are not entirely helpful, because browsers are limited to testing the serialization results of CSS.

**Comment by ai** [comment]
I recommend copying `indexOf`/RegExps jumps first since they are bringing the main performance improvements

https://github.com/postcss/postcss/blob/master/lib/tokenize.es6#L253-L254
https://github.com/postcss/postcss/blob/master/lib/tokenize.es6#L161-L175
https://github.com/postcss/postcss/blob/master/lib/tokenize.es6#L116
https://github.com/postcss/postcss/blob/master/lib/tokenize.es6#L284-L285

JS is very slow when you are tokenizing symbol by symbol.

**Comment by jonathantneal** [comment]
I’m feeling much better after analyzing those jumps. Let me share what I’ve learned. I’d like to know what you think.

I was able to implement the comment jump. There was no perceivable yield. This is what it looked like:

```js
			// comment
			case FS:
				if (css.charCodeAt(pos + 1) === STAR) {
					typ = 'comment'
					pos = css.indexOf('*/', pos + 2) + 1
					if (pos === 0) {
						pos = len
					} else {
						++pos
					}
				} else {
					++pos
				}
				break
```

I was able to implement the string jump. There was no perceivable improvement. This is what it looked like:

```js
			// string
			case DBLQ:
			case SNGQ:
				var quote = cp0 === DBLQ ? '"' : "'"
				var escaped
				do {
					escaped = false
					pos = css.indexOf(quote, pos + 1) + 1
					if (pos === 0) {
						pos = len
						break
					}
					escapePos = pos - 1
					while (css.charCodeAt(escapePos - 1) === BS) {
						escapePos -= 1
						escaped = !escaped
					}
				} while (escaped)
				break
```

So far, it seems those jumps are not beating symbol checks.

However, after looking at the next two jumps — I think I see where some of the performance differences are coming from. However, I’m not sure I should implement those jumps.

I did not want to implement the parentheses jumps because they skip tokenizing the contents between the parentheses. This is probably one of the reasons the current tokenizer is picking up 49,548 tokens from `bootstrap.css`, while this experimental tokenizer is picking up 60,458 tokens.

Put another way; the experimental tokenizer is 1.4 times slower but it is also picking up 1.2 times more tokens, and it is accurately handling the escapes.

---

Up Next: I think I have time tonight to wire in this new tokenizer to the current object model.

**Comment by ai** [comment]
> I did not want to implement the parentheses jumps because they skip tokenizing the contents between the parentheses. This is probably one of the reasons the current tokenizer is picking up 49,548 tokens from bootstrap.css, while this experimental tokenizer is picking up 60,458 tokens.

Yeap, we need to count PostCSS core + PostCSS Value parser time.

If new tokenizer will preare tokens for value/selector parsing, a little slow down will bring a big speed up in value/selector parsing.

**Comment by jonathantneal** [comment]
I haven’t had much time, but I have added another test that runs PostCSS with Selector and Value parsing. The functionality looks like this, which I hope we’ll find generous:

```js
parsed.root.walk(node => {
  switch (node.type) {
    case 'decl':
      postcssValuesParser(node.value)
      break
    case 'rule':
      postcssSelectorParser.processSync(node.selector)
      break
  }
})
```

In reality, I think most plugins would runs those parsers in separate walks, and most plugin packs would run those parsers multiple times. Still, here are the updated parse times:

```
Collecting PostCSS Parser Benchmarks...

PostCSS Parser 7.0.29:              1204 nodes in 14 ms
PostCSS Experimental Parser:       13686 nodes in 17 ms (1.2 times slower)
PostCSS + Selector + Value Parser:  1204 nodes in 78 ms (5.5 times slower)
```

**Comment by jonathantneal** [comment]
Andrey, please let me know if I’m pestering with these updates or if I can make them more helpful.

While analyzing the "slower" parts of the tokenizer, it seems like eagerly checking the character ahead improves overall performance. I have rewritten the tokenizer to take advantage of this. I have also added 2 fields to a token; they are the opening and closing distances between the meaningful value of a token and its delimiters. Although “_delimiter_” is a poor term, this refers to the split between things like the `@` & `media` in a `@media` At-Identifier token, the `2` & `em` in a `2em` Number token, or the `"` & `hello` & `"` in a `"hello"` String token.

Anyway, I think you’ll really like these results!

```
Compressing PostCSS Tokenizer...

PostCSS Tokenizer Development:       1910 B
PostCSS Tokenizer Development (min):  638 B
PostCSS Tokenizer Development (web):  639 B

Collecting PostCSS Tokenizer Benchmarks...

PostCSS Tokenizer Development:       58721 tokens in 8 ms (1.0 times faster)
PostCSS Tokenizer Development (min): 58721 tokens in 8 ms (1.0 times faster)
PostCSS Tokenizer 7.0.30:            49548 tokens in 8 ms


Compressing PostCSS Parser...

PostCSS Parser Development:       1369 B
PostCSS Parser Development (min):  836 B
PostCSS Parser Development (web):  805 B

Collecting PostCSS Parser Benchmarks...

PostCSS Experimental Parser:       56024 nodes in 10 ms (1.6 times faster)
PostCSS Parser 7.0.30:              6240 nodes in 15 ms
PostCSS + Selector + Value Parser: 28491 nodes in 86 ms (5.5 times slower)
```
— From https://github.com/csstools/tokenizer#collecting-postcss-parser-benchmarks

**Comment by ai** [comment]
> Andrey, please let me know if I’m pestering with these updates or if I can make them more helpful.

A new performance breakthrough is awesome 😍.

Let’s change tokenizer in 8.0.

Is it possible to use it in the safe parser or SCSS parser without changes? I forked the current one and can fork a new one too.

**Comment by evilebottnawi** [comment]
@ai what is status this feature for 8 version? Postpone?

**Comment by ai** [comment]
@evilebottnawi here is what we need to do before the release
https://github.com/postcss/postcss/milestone/5

You can track the progress here
https://twitter.com/PostCSS/status/1298332535525695491

**Comment by ai** [comment]
@evilebottnawi yeap, we decided to do it after 8.0, since new parser will make 8.0 release too big and will force us to wait too long

**Comment by evilebottnawi** [comment]
Sounds good :+1:

**Comment by romainmenke** [comment]
Took me a while to analyse this issue because initially I interpreted this as "PostCSS has a bug with escaped symbols".

But this is not a bug in PostCSS. As far as I can tell PostCSS produces the correct AST with or without escaped symbols.

It produces a few words too many at the tokenizer level, but this goes away in the parser.

The original issue was that a downstream package did not produce the correct AST and uses the same tokenizer. Changing the tokenizer here would produce a different output there and might have fixed some issues.

From a different perspective this sounds very risky to me.
A fix for one issue might be a number of new bugs for other cases.

That the downstream package uses internals from PostCSS also caused other issues and most plugins are no longer uses this dependency because of that.

-----

I think that this has a lot of overlap with : https://github.com/postcss/postcss/issues/1145#issuecomment-1397999466
